### ISSUE TITLE
[2020 ICLR] Decoupling Representation and Classifier for Long-Tailed Recognition

### ISSUE THREAD

**Issue by Jasonlee1995** [Jasonlee1995/AI_Papers#80]
Long-tail distribution에서 발생되는 class imbalance 문제를 해결해오던 loss re-weighting, data re-sampling, transfer learning 등과 같은 방법들은 representation과 classification을 같이 학습하였음

해당 논문은 representation learning과 classification 과정을 decouple하여 class imbalance 문제를 해결해오던 방식이 어떠한 영향을 주는지 분석하고 이로부터 간단한 방식을 제시함

핵심은 instance-balanced sampling으로 representation을 학습하고 classifier만 조절하여도 충분히 잘 classification한다는 것
(이러한 결론에 도달하게 된 배경은 classification layer의 weight norm이 imbalance 정도인 class cardinality와 correlation이 되어있다는 관측 현상 때문)

논문에서 말하는 main contribution은 다음과 같음

```
1. Instance-balanced sampling learns the best and most generalizable representations
2. Retraining the classifier with class-balanced sampling or classifier weight normalization is enough
3. Achieve state-of-the-art on multiple longtailed recognition benchmark datasets
```

## 1. Learning Representations for Long-Tailed Recognition

Long-tailed recognition에서 training set은 long-tailed distribution을 따라가며, 이러한 imbalanced dataset으로 학습된 모델은 few-shot classes에 대해 under-fitting이 발생함

기존의 re-sampling strategies, loss reweighting, margin regularization 등의 방식을 통해 성능을 끌어올릴 수 있었음

그러나 이러한 방법들이 어떻게 성능을 발전시켰는지에 대해선 명확하지 않음

저자들은 representation learning과 classification learning을 disentangle하여 해당 방법들이 문제를 어떻게 해결했는지 분석함

### 1.1. Sampling strategies
<img width="1079" alt="image" src="https://user-images.githubusercontent.com/49643709/181962761-476145a7-b4b6-409a-9a35-f02f1e8ae04a.png">
<img width="1043" alt="image" src="https://user-images.githubusercontent.com/49643709/181692480-a819a8bf-000b-4c9a-a5a2-5b1a7391cebb.png">

#### 1.1.1. Instance-balanced sampling

`q = 1`인 경우로 각 데이터가 선택될 확률이 동일
(일반적으로 수행하는 sampling 방법)

#### 1.1.2. Class-balanced sampling

`q = 0`인 경우로 각 class가 선택될 확률이 동일

Class를 uniform하게 sample한 후, 해당 class에 해당하는 데이터를 uniform하게 sample하는 2-stage sampling strategy로 생각할 수 있음

#### 1.1.3. Square-root sampling

`q = 0.5`인 경우

#### 1.1.4. Progressively-balanced sampling
<img width="1042" alt="image" src="https://user-images.githubusercontent.com/49643709/181699545-ddad8b0e-f170-4ea9-9cf8-27c0a79ebdb0.png">

처음에는 instance-balanced sampling으로 학습하다 시간이 지나면 지날수록 class-balanced sampling이 되도록 scheduling

### 1.2. Loss re-weighting strategies

대부분의 high performance를 찍은 recent approaches는 학습하기 어려우며 reproduce가 잘 안됨

그리고 많은 경우 extensive, dataset-specific hyper-parameter tuning을 요구함

따라서 이와 관련된 모든 approaches를 확인하지 않음

## 2. Classification for Long-Tailed Recognition
<img width="980" alt="image" src="https://user-images.githubusercontent.com/49643709/181919426-b463ea10-87c3-4e14-942e-ef99edda8232.png">

일반적인 classification task에서 representation과 classification을 같이 학습함 (train jointly)

Long-tailed recognition의 다양한 approaches 또한 representation과 classifier를 같이 학습함

저자들은 representation과 classification을 decoupling하여 classifier를 학습시키는 방식을 여러가지 제안함

### 2.1. Classifier Re-training (cRT)

Representation은 고정한 채 classifier만 class-balanced sampling으로 학습

Classifier는 randomly re-initialize하며 class-balanced sampling을 이용하여 적은 epoch만큼 학습

### 2.2. Nearest Class Mean classifier (NCM)

각 class에 해당하는 mean feature representation을 구한 후 nearest neighbor search를 통해 classification

Cosine similarity 혹은 L2 normalized mean feature과의 Euclidean distance를 통해 nearest neighbor을 찾음

### 2.3. $\tau$-normalized classifier ($\tau$-normalized)
<img width="400" alt="image" src="https://user-images.githubusercontent.com/49643709/181918834-31628742-c4d1-4706-ae03-784e74e5aeae.png">

저자들은 instance-balanced sampling을 통해 학습한 모델의 classification layer weight의 norm이 class의 cardinality와 correlation이 있으며, class-balanced sampling을 통해 fine-tune한 classifier의 weight norm은 서로 더 비슷함을 관측함

관측된 현상을 통해 단순히 classifier의 weight norm을 $\tau$-normalization procedure을 통해 조절하여 decision boundary의 imbalance를 조정

<img width="977" alt="image" src="https://user-images.githubusercontent.com/49643709/181918979-562ad752-3931-4594-9778-b0248e20a761.png">

$\tau$는 normalization의 temperature를 조절하는 hyperparameter로 empirically하게 0과 1 사이에서 선택
(bias term은 제거, $\tau$가 1인 경우 L2 normalization과 같으며 0인 경우는 original weight과 같음)

### 2.4. Learnable Weight Scaling (LWS)
<img width="375" alt="image" src="https://user-images.githubusercontent.com/49643709/181919124-be177c57-a5c0-4b12-ad73-97060f47ec73.png">

$\tau$ normalization을 classifier weight의 방향은 유지한 채 magnitude만 re-scale한 것이라고 볼 수 있음

$\tau$는 cross-validation을 통해 구하는 데, 위 식에서의 $f_{i}$를 training set의 class-balanced sampling을 통해 학습

즉 cRT와 비슷하게 representation과 classifier을 학습한 후, representation과 classifier weight를 고정해두고 scaling factor $f_{i}$만 학습을 통해 구함

#### 3.1.1. Dataset

Places-LT, ImageNet-LT, iNaturalist 2018 dataset 사용

- Places-LT : 365 categories with images per class ranges from 4980 to 5
- ImageNet-LT : 1000 categories with images per class ranges from 1280 to 5
- iNaturalist 2018 : 8142 categories (real-world, naturally long-tailed dataset)

Places-LT, ImageNet-LT는 balanced version인 Places-2, ImageNet-2012 dataset을 인위적으로 long-tailed distribution이 되도록 artificially truncated해서 만든 large-scale long-tailed dataset

#### 3.1.2. Evaluation Protocol

Balanced test / validation dataset에서 model evaluation

추가적으로 many-shot (> 100 images), medium-shot (20 ~ 100 images), few-shot (< 20 images) classes에 대해 accuracy report

#### 3.1.3. Implementation

- Batch size 512
- SGD optimizer with 0.9 momentum
- Cosine learning rate schedule gradually decaying 0.2 to 0
- Input image resolution : 224 x 224
- Training epochs : 90

### 3.2. Sampling Strategies and Decoupling Learning
<img width="1083" alt="image" src="https://user-images.githubusercontent.com/49643709/181936284-e606d211-642c-458a-8323-ef5c3058d8ac.png">

#### 3.2.1. Sampling matters when training jointly
<img width="1092" alt="image" src="https://user-images.githubusercontent.com/49643709/181936395-79b782f7-e7a2-44a5-b587-d6aa6e0815c5.png">

Jointly training을 할 경우, better sampling strategies를 사용할 수록 performance는 consistent하게 오름

Many-shot classes에서 instance-balanced sampling이 highest performance를 보이는 것을 관측할 수 있는데, 이는 instance-balanced sampling 방법으로 학습한 모델은 many-shot classes에 대해 highly skewed 되어있음을 알 수 있음

#### 3.2.2. Joint or decoupled learning?
<img width="538" alt="image" src="https://user-images.githubusercontent.com/49643709/181936892-6c80da48-7942-4b34-b777-5b3befaa7138.png">

Representation과 classifier를 decouple하면 beneficial함을 확인할 수 있음

#### 3.2.3. Instance-balanced sampling gives the most generalizable representations

`Figure 1`을 통해 instance-balanced sampling을 통해 학습한 모델에 decoupling method를 적용한 경우가 성능이 제일 좋음을 확인할 수 있음

이는 매우 흥미로운 현상인데, data imbalance 문제는 high-quality representation을 배우는 데 issue가 되지 않음을 알 수 있음

### 3.3. How to Balance your Classifier?
<img width="1073" alt="image" src="https://user-images.githubusercontent.com/49643709/181940123-d300fbb5-543d-4328-ae30-57de9eeb532c.png">

`Figure 2`를 보면 instance-balanced sampling으로 학습한 모델의 경우, classifier의 weight vector의 L2 norm이 class training images와 positively correlated되어 있음을 알 수 있음
(more-shot classes tent to learn a classifier with larger magnitudes)

Weight scaling hyperparameter $\tau$를 조절하여 성능 평가 실험한 결과 weight norm의 값을 balancing해주면 many class의 accuracy는 감소하나 medium, few class accuracy는 증가하여 total accuracy는 증가함을 알 수 있음

즉 기존의 instance-balanced sampling으로 모델을 학습하면 more-shot class의 decision boundary를 wider하게 만들어 data-scarce classes의 성능을 해치고 data-rich classes는 성능을 높이도록 학습됨

Decoupling해서 classifier의 weight의 방향을 그대로 둔 상태에서 scale을 줄여 decision boundary의 wide를 어느정도 balancing하여 성능이 올라간다고 생각하면 됨

### 3.4. Comparison with the SOTA on Long-Tailed Datasets

<img width="1078" alt="image" src="https://user-images.githubusercontent.com/49643709/181958148-ce10fdd3-121e-4fc9-9811-81a2c629d0be.png">
<img width="1091" alt="image" src="https://user-images.githubusercontent.com/49643709/181959462-ea78dbd4-9623-4e83-9393-bb33a32278ff.png">

**Comment by Jasonlee1995** [comment]
Appendix에 생각할만한 점들이 있어 추가 + 생각들에 대한 메모

### Classifier Decision Boundary

Feature가 고정된 상태에서 linear classifier을 학습한다고 하면 어떻게 학습이 될지에 대한 그림을 그려보면 다음과 같음

<img width="470" alt="image" src="https://user-images.githubusercontent.com/49643709/181966458-14b3d11f-fcdc-47a6-a96c-2031b2cae9c0.png">
<img width="534" alt="image" src="https://user-images.githubusercontent.com/49643709/181966853-8fd61482-f661-4877-84b1-334105c8e295.png">

Classifier에 해당하는 weight matrix와 feature vector간의 linear product가 됨을 생각해보면 다음과 같이 생각할 수 있음
(특정 class에 해당하는 weight vector는 D x 1 shape의 vector임을 알 수 있음)

이로부터 논문에서 제시한 decoupling method가 어떻게 작동하는 지에 대해 `Figure 4`를 보면 더욱 직관적으로 이해할 수 있음

<img width="1096" alt="image" src="https://user-images.githubusercontent.com/49643709/181964580-78245c11-3a3f-41b2-bc0f-5066077772bb.png">


### On the Exploration of Determining $\tau$
<img width="783" alt="image" src="https://user-images.githubusercontent.com/49643709/181977730-60530cd8-fa92-4bd2-abfe-9f9e94b2baae.png">

적합한 $\tau$를 찾는 것 자체가 많은 practical scenario에서 disadvantage가 됨

저자들은 이를 위해 train dataset에서 적합한 $\tau$를 찾는 2가지 방법을 제안하며 이는 효과 있음을 실험을 통해 보임

1. Simulating a balanced testing distribution from the training set
2. Learning $\tau$ value on training set (LWS와 다른 점은 hyperparameter가 $\tau$ 하나인지 아니면 class마다 있는지임)


### Comparing MLP Classifier with Linear Classifier
<img width="826" alt="image" src="https://user-images.githubusercontent.com/49643709/181994277-86ac954b-8d9c-4414-9a6f-a501a996079f.png">
<img width="508" alt="image" src="https://user-images.githubusercontent.com/49643709/181994438-26c4ffb8-30b0-4f03-a516-e5fc5f5ebc64.png">

Representation을 fix한 상태에서 balanced sampling을 이용하여 classifier 학습

MLP layer가 deep해질수록 성능이 저하되는 현상이 관측되며, 이는 backbone network가 충분히 discriminative representation을 배웠음을 의미함


### Cosine Similarity for Classification
<img width="548" alt="image" src="https://user-images.githubusercontent.com/49643709/181994446-15fa8b26-8fd7-486e-9f9d-697c764856c1.png">

다른 분야에서 linear classifier 대신 cosine similarity classifier를 사용하는 사례가 존재했음

Linear classifier 대신 cosine similarity classifier를 사용하여 실험 결과, 충분히 comparable함을 알 수 있음

논문에서 제시한 방법과 다른 점은 classification weight L2 normalize해주고 feature vector까지 L2 normalize해준다는 점 정도

<img width="666" alt="image" src="https://user-images.githubusercontent.com/49643709/181994518-cb4e3da6-a00c-4e11-b286-4fbeb60f21d3.png">
